### ISSUE TITLE
Method to request permission

### ISSUE THREAD

**Issue by ClaesNilsson** [sysapps/tcp-udp-sockets#81]
This API is too sensitive to be exposed to any arbitrary web site and must thus be implemented together with a security and permission model. Work in W3C on “Trust and Permissions for Web Applications” is starting up as there currently is no standardized model to achieve secure access to sensitive APIs. So in order to expose this API to web runtimes we have so far to rely on platform dependent solutions to achieve secure access. I don't think that access control for this API could be based on user consent only so a solution based on existing standardized security mechanisms such as tls/ssl, signed manifests, csp, etc is probably needed.   

To be able to focus on the API design itself, and hide the details of the API access/permission system, it is proposed that a permission request method for each socket type is added. The method returns a promise that is, if permission was given through the platform’s security system, resolved with the created socket object as parameter.

It is proposed that the following replaces the current UDP, TCP and TCPServer object constructors:

```
partial interface Navigator {
    readonly    attribute UDP udp;
    readonly    attribute TCP tcp;
    readonly    attribute TCPServer tcpServer;
};

Interface TCP {
    Promise<TCPSocket> get (DOMString remoteAddress, unsigned short remotePort, 
                                            optional TCPOptions options);
};

Methods:

get

If permission is given this method creates a TCPSocket object and starts a TCP connection
attempt  with the remote host. The permission system is implementation dependent. 

This following steps MUST run::

1. Create a new Promise, getTCPPromise, return it and run the remaining steps asynchronously.
2. Check if the client web application has permission to access this API and connect with the 
     requested remote host. If the client web application does not have permission to create a  
    TCPSocket object then reject getTCPPromise with SecurityError (not sure on error name here), 
    else create a new TCPSocket object ("mySocket"). 

Step 2- 11 in the stepwise description of the TCPSocket constructor semantics in the current 
version of the specification.

13. Resolve getTCPPromise with the mySocket as argument.

Example:

navigator.tcp.get("127.0.0.1", 6789).then(
    tcpSocket => {
       //  Use the socket …..
    },
     e => {
         // Handle “not permission given”
    }
);
```

Similar for UDP and TCPserver.

An alternative could be to extend the proposed Permission API, http://lists.w3.org/Archives/Public/public-webapps/2014JulSep/0389.html, with a method to request permission to access APIs.

WDYT?

**Comment by domenic** [comment]
I don't like the idea of replacing the constructors since then how do the objects end up being created anyway? You can try to pretend there is a constructor that only the UA can use, but that is very magical.

Instead I would just have appropriate permission-asking method(s), and if you try to construct instances before permission is granted, the constructors immediately throw.

**Comment by ClaesNilsson** [comment]
Thanks for your reply Domenic. I see your point with the UA’s “magic constructor”. 

I had a look at the [W3C Push API](https://w3c.github.io/push-api/) and consider something similar:

```
partial interface Navigator {
    readonly    attribute UDPPermission udpPermission;
    readonly    attribute TCPPermission tcpPermission;
    readonly    attribute TCPServerPermission tcpServerPermission;
};

interface TCPPermission {
    Promise<TCPUDPPermissionStatus> hasPermission (DOMString remoteAddress, 
                  unsigned short remotePort);
    Promise<void>  requestPermission (DOMString remoteAddress, unsigned short remotePort);
};

Methods:

hasPermission

Retrieves the permission status (TCPUDPPermissionStatus) of the requesting webapp for creating
a TCPSocket object. This method  returns a Promise with permission status “granted”, “denied” or
“default” (the webapp has to request  permission). 

requestPermission

Requests permission to create a TCPSocket to this specific remoteAddress and remotePort.
Permission can be based on user consent, based on trust for this webapp (.e.g. based on web
security mechanisms such as secure transport, signature and CSP) or previously granted or 
denied permission explicitly for this webapp. The method returns a Promise, which is resolved if
permission was given and rejected if permission was denied. 

Example:
navigator.tcpPermission.hasPermission("127.0.0.1", 6789).then(
    permissionStatus => {
       if (permissionStatus === “default”) {
           //  Request permission …..
           navigator.tcpPermission.requestPermission("127.0.0.1", 6789).then(
                () => {
                     // Create and use TCPSocket object
                     var mySocket = new TCPSocket("127.0.0.1", 6789);
                      ………………
```

This means that the webapp has to perform three steps before the TCP socket object can be used. On the other hand webapps have more choices based on permission status.

Something similar could be done for UDP and TCPServer sockets but the remote peer’s address and port may not be known prior to creating the socket.

WDYT?

**Comment by domenic** [comment]
In general this sounds good. You may be able to tie in to https://w3c.github.io/permissions/ to avoid adding so many new properties and methods.

> This means that the webapp has to perform three steps before the TCP socket object can be used.

They should be able to just call requestPermission without calling hasPermission first; if the permission is already granted this will be a no-op.

**Comment by ClaesNilsson** [comment]
I was aware on the work with the Permissions API but had not seen this latest, though still "unofficial", draft. Yes, of course the best approach for the TCP and UDP Socket API is to tie this into a general Permissions system if this is standardized.

So if we assume that the general [Permissions API](https://w3c.github.io/permissions/) can be used to give information on the permission state then only a requestPermission method has to be added to the TCP and UDP Socket API. 

Looking at the [Permissions API](https://w3c.github.io/permissions/) new PermissionNames “udp”, “tcp” and “tcpserver has to be added. In addition a dictionary TCPPermission, inheriting from dictionary Permission, adding fields for remote address and port has to be defined. 

Also note my comment on Alex Russel’s post  [A Plan For Permission-Requesting APIs](https://gist.github.com/slightlyoff/43cd8c2f64a0719358fe).

**Comment by domenic** [comment]
We should tag in @mounirlamouri @marcoscaceres and @slightlyoff for their thoughts. It would be nice to get a request() method added to the permissions API so that you don't have to spin your own.

**Comment by ClaesNilsson** [comment]
@mounirlamouri @marcoscaceres and @slightlyoff , your views are appreciated.

**Comment by marcoscaceres** [comment]
Yes, we can add this to the permission API, but only once we have commitments to actually implementation this to the platform. However, the timing for this API is not great or high priority for us at Moz. The permission problems come mostly from the UX side, not so much from the technical side.

**Comment by ClaesNilsson** [comment]
Ok, thanks Marcos. Based on what you say and the fact that the Permission API is an unofficial draft I guess that I for now have to add the two methods proposed in https://github.com/sysapps/tcp-udp-sockets/issues/81#issuecomment-65085352 and make a note that they will be replaced by a general solution if that is standardized.

**Comment by marcoscaceres** [comment]
That seems fine, but please note that the "official" status of specifications is not really relevant to the discussion. What matters is who is implementing the specs. Both Google and Mozilla are interested in standardizing and implementing the Permissions API. However, if you want to wait until we publish it somewhere , that's ok too.

**Comment by slightlyoff** [comment]
Regardless, the Permissions API has many unresolved issues as yet. The API
should, at minimum, use an Asynchronous Request API Pattern
https://gist.github.com/slightlyoff/43cd8c2f64a0719358fe and integrate
with a Permissions API if/when it's available.

In general, it's an anti-pattern to lengthen the critical path for any spec.

Regards

**Comment by ClaesNilsson** [comment]
Ok, thanks for your inputs Domenic, Marcos and Alex. The major remaining issue with the current version of the TCP and UDP Socket API specification is that it is very vague on the security model. Even though work is ongoing with the general [Permissions API](https://w3c.github.io/permissions/) I suggest that I add something like what I am proposing in https://github.com/sysapps/tcp-udp-sockets/issues/81#issuecomment-65085352 and a comment that this could later be replaced by a general solution, for example like the [Permissions API](https://w3c.github.io/permissions/) extended with an Asynchronous Request API Pattern if/when that is standardized and gets momentum.

This should hopefully clarify our thinking on this major issue for this API and could also provide input to the general discussion on trust and permissions for sensitive APIs.

**Comment by ClaesNilsson** [comment]
@domenic, @mounirlamouri, @marcoscaceres, @slightlyoff 
I have issued PR https://github.com/sysapps/tcp-udp-sockets/pull/82.

So far I have only defined permission methods for TCP as I would like to get feedback before doing something similar for UDP and TCPServer.

**Comment by marcoscaceres** [comment]
Again, this isn't a JS API issue: it's a UX/UI issue (at least given the current Web security model). It's true that, at least from the spec's perspective, all we can really do is add `.requestWhatever()`, but if we don't work out how end-users would make sense of this feature, then this is a nonstarter. 

I've discussed this with folks at Mozilla and feel fairly strongly that we need to go back to the use cases we are actually trying to address with this API. "Talking to the Internet/Intranet" seems overly broad. So what are we trying to do here?

**Comment by ClaesNilsson** [comment]
Thanks Marcos,

Going back a little SysApps was originally chartered to provide a runtime and security model so that it would be possible to open up sensitive APIs to SysApps enabled runtimes. When I started to edit the TCP and UDP Socket API the existence of such a security model was assumed. In addition, looking at existing TCP and UDP Socket APIs they are implemented in proprietary web runtimes, FFOS and Chrome, which provides a security model for installed packaged web runtimes. 

However, today we can conclude that it has not been possible to standardize a runtime and security model in SysApps. We can also conclude that there still is a strong interest in a standardized raw UDP and TCP socket API, which has been expressed for example by Google and Mozilla. So the question remains, for which kind of web runtime could this API be opened up to and how does the security/permission model look like? Questions:
- Should this API only be exposed in web runtimes for packaged installed web apps such as FFOS, Chrome Apps and Tizen? 
- Could this API, or a subset of it, be exposed to server hosted web apps running in the normal web browser?  Personally I doubt that it is possible to rely on user consent only for this API but a combination of user consent and web security mechanisms such as https:, signed manifest and maybe certificate pinning may be a way forward. Sony has an interest in a security/permission model for sensitive APIs exposed for server hosted web apps. See for example our project with [FFOS Trusted Hosted Web Apps](http://lists.w3.org/Archives/Public/public-sysapps/2014Sep/att-0000/SoMC_FFOS_Trusted_Hosted_Apps.pdf), mainly slide 7. Could something similar be done for a web browser?

Currently the TCP and UDP Socket API may be underspecified when it comes the security/permission, but as there currently exists no standardized model for how to expose sensitive APIs to either installed packaged web apps or to server hosted browser web apps, the ambition with the requestPermission method was to “decouple” the security/permission model from the API specification itself. That is, the security/permission model depends on the type of web runtime the API is implemented in.

Coming back to the question on use cases (sorry for this lengthy introduction Marcos), our interest in a standardized raw UDP and TCP socket API started with us seeing the need for implementing local network multicast service discovery protocols and the need to communicate with services on local devices top of UDP and TCP. Our assumption was a trusted web execution environment relieving users from informed decisions. 

My interpretation is that there still is an interest this API and that it should be exposed to server hosted web apps (please correct me if I a wrong). This might be controversial, consider for example that the [Network Service Discovery API](http://www.w3.org/TR/2014/WD-discovery-api-20140220/), which is an API aimed for the browser environment, has not yet been implemented due to security issues. However, I am extremely happy to receive feedback with ideas on how we could find a security/permission model for the TCP and UDP Socket API. Questions are for example:
- In addition to the use cases I mention above, which kind use cases should we support?
- To which level should the security/permission model be defined in the API specification? Is the current approach with “requestPermission” and leaving the manner in which permission is given up to the API implementation and type of web runtime enough?
- To which degree could we rely on user interaction and to which degree do we need to rely on web security mechanisms such as https: (see discussion on requiring https: for Geolocation API), signed manifest and certificate pinning?

**Comment by marcoscaceres** [comment]
Thanks @ClaesNilsson for the detailed response. I'll try to address some of the points. 

> Should this API only be exposed in web runtimes for packaged installed web apps such as FFOS, Chrome Apps and Tizen? 

If the answer to this is "yes", then standardizing this at the W3C would be inappropriate as it would have limited value to the Web. SysApps is effectively dead (in all but name), so there is no avenue to actually do this standardization right now.   

> Could this API, or a subset of it, be exposed to server hosted web apps running in the normal web browser?

Depends on the problem that is being solved... 

> Sony has an interest in a security/permission model for sensitive APIs exposed for server hosted web apps. See for example our project with FFOS Trusted Hosted Web Apps, mainly slide 7. Could something similar be done for a web browser?

Maybe, but this requires a lot of exploration. Previous attempts to deviate unnecessarily from the web security model have failed (I would include "hosted apps" in this bucket) - so it's best to work with the Web. 

Fortunately, there is work going on in this area (though I'm still not particularly familiar with what is being proposed):
https://w3c.github.io/webappsec/specs/powerfulfeatures/ 

>  the ambition with the requestPermission method was to “decouple” the security/permission model from the API specification itself. 

That's fine and "the right thing to do" :tm:, but again - that's just side-stepping the problem.

> This might be controversial, consider for example that the Network Service Discovery API, which is an API aimed for the browser environment, has not yet been implemented due to security issues. 

It's not really controversial. This API will hit the same roadblocks. It's also not just security issues - failure to engage browser vendors has also been detrimental to the adoption of that API. 

> ```
> In addition to the use cases I mention above, which kind use cases should we support?
> ```

As this is outside my domain of knowledge, the use case: "implementing local network multicast service discovery protocols and the need to communicate with services on local devices top of UDP and TCP" is very vague to me. 

What specific protocols, products, services, devices are you trying to communicate with exactly? And, I guess, why is the Network Service Discovery API insufficient (irrespective of implementer interest)? 

AFAIK, Mozilla's primary use case for in this API was to talk to IMAP servers directly. 

At a minimum, I would expect a list of user stories. It also depends on the limitations of the platform to achieve those things. 

>  To which level should the security/permission model be defined in the API specification?

That really depends on what you are trying to do. 

>  Is the current approach with “requestPermission” and leaving the manner in which permission is given up to the API implementation and type of web runtime enough?

Sometimes, yes.   

>  To which degree could we rely on user interaction and to which degree do we need to rely on web security mechanisms such as https: (see discussion on requiring https: for Geolocation API), signed manifest and certificate pinning?

Again, it depends on what one is trying to do. 

Please also note that there has not been any discussion of "signed manifests" at the W3C (at least, not one has discussed them with me).

**Comment by ClaesNilsson** [comment]
Thanks again Marcos @marcoscaceres,

We have some major questions that partly go beyond this specific specification.
You state:

> SysApps is effectively dead (in all but name), so there is no avenue to actually do this standardization right now.

At last TPAC it was said that the plan is thatSysApps should be re-chartered. However, it has not yet happened and the activity on the mailing list is almost null. So we need a clear statement on the future of this WG and the deliverables of it. If I remember correctly from the meeting there was interest in at least the Task Scheduler API and the TCP and UDP Socket API. So if SysApps is closed new homes for those specifications have to be defined. 

Regarding use cases we have earlier conducted research projects in which we implemented JS-libraries for UPnP/SSDP using legacy socket APIs, e.g. Chrome sockets. Compared to using the Network Service Discovery API we saw that as a more flexible solution as we assumed a standardized TCP and UDP socket API running in a secure web runtime (as was chartered by W3C SysApps).  We also have other web projects where a socket API is used but I need to investigate what we could disclose here.

Looking at where we are today we need to make some decisions. Your view that APIs requiring a secure runtime for installed webapps is out of scope for standardization may not be shared by everyone. 

However, personally I like the idea to evolve the web so that powerful APIs could be exposed in the “normal web browser” and https://w3c.github.io/webappsec/specs/powerfulfeatures/ seems like an interesting step. 

I have heard different people express their interest in a standardized TCP and UDP Socket API and I am curious on more views on use cases and the security/permission model. For example, what does Google say, @domenic, @mounirlamouri, @slightlyoff, @jyasskin.

**Comment by jyasskin** [comment]
@adrifelt too. I know we're interested in finding ways to expose TCP and UDP sockets to web pages, but so far I think we've gotten stuck on the "user consent" step. How do you explain to the user what "ldap.corp.com:636" means in a way that lets them decide whether to allow access to it, especially after the app has given a potentially-misleading explanation of the same thing? As Marcos said, it's good to use the `requestPermission` model to let different UAs ask the user in different ways, but that doesn't manage to "decouple" the API from the permission story. In order to ship the spec, you need implementations, and in order to get browser implementations, the browsers do have to figure out the permission story. It'd be helpful for this repository to give that story, whether or not it winds up in the formal specification. 

I believe @slightlyoff wants to follow the Chrome Apps model where we'd give up on giving the user an explanation they can understand, and instead rely on a central app store to inspect packages (http://w3ctag.github.io/packaging-on-the-web/) before signing them. This allows the store to centrally revoke the permission for a package to run if it's discovered to be malicious. I'm unhappy with that model because of the centralization and because app stores haven't been able to get apps to follow users' wishes in the past, but it is plausible if the use cases are compelling enough.

**Comment by marcoscaceres** [comment]
I agree, it would be sad if we centralized. There is no guarantee that a Chrome Store signed package would run in, say, Firefox, Safari, or IE or any new browser that might want to come along - or vice versa. It would also make it extremely hard for new app stores to appear without the blessing of browser makers. 

We've seen this already with things like Firefox OS, where the promise was to allow multiple marketplaces to bloom, but then any powerful features could only be enabled by going through the Firefox Marketplace (and getting Mozilla's digital signature of approval). As with Chrome Apps, going through a centralized marketplace effectively kills the network effect of the Web and discourages API interoperability (again, Chrome Apps and FxOS as examples).

**Comment by adrifelt** [comment]
What are the main security and privacy risks of the API? The current draft doesn't say.
- Websites won't have access to the cookie store, so access to HTTP resources isn't a concern.
- But access to a local network might be a security issue.

So the question we would be asking the user is, "Can this website be trusted with access to your local network?" The problem I see is that users aren't qualified to answer that question. Even a very tech-savvy person won't necessarily know what's available on their local network. Only the network admin can answer that question. Would it instead be possible to put in technical limitations that prevent this class of threat? For example, @jyasskin pointed me to http://www.w3.org/TR/discovery-api/#security-considerations-for-api-implementations which has network admins define whether or not they support CORS.

**Comment by ClaesNilsson** [comment]
Thanks, this is a good discussion. 

Existing implementation efforts that refers to this specification that I am aware of are for installed packaged web apps, [FFOS UDP](https://bugzilla.mozilla.org/show_bug.cgi?id=745283) and [an experimental implementation for Crosswalk](https://crosswalk-project.org/documentation/apis/web_apis.html).

If we go beyond the original assumption of a secure installed “SysApps” web app model for providing a W3C standardized TCP and UDP Socket API we have to reconsider what we want to achieve and what could achieve in a secure manner. As said before, I doubt that we could find a model based on user consent only for this kind of sensitive API. Requiring https: transport is of course necessary but not enough. I still think we should to find a model using existing, and evolving, web security mechanism so that this, and other sensitive APIs, could be opened up for web server hosted apps.   

So far my efforts on this API have been spent on solving the UDP and TCP communication and a lot of efforts were spent on the adaptation of the API to the WHAT WG Streams API. If there is a true interest by browser vendors to implement this API I am happy to continue to act as an editor of the API and together work on solving the security/permission issues. As this goes beyond the original assumptions, i.e. a “sysapps environment”, I am happy for support by people skilled in security/permission issues.

As a starting point I suggest that I complete the PR https://github.com/sysapps/tcp-udp-sockets/pull/82 with permission methods for UDP and TCPServer as well. 

WDYT? @domenic, @slightlyoff, @jyasskin, @adrifelt, @marcoscaceres

**Comment by jyasskin** [comment]
Completing #82 won't be sufficient for browsers to implement the API, but it's also a step forward so +1 to completing and merging it.

**Comment by marcoscaceres** [comment]
Agree. But after that I would suggest putting this API on a holding pattern till we work out the actual use cases for it (don't want to end up with a solution looking for a problem). 

Given what has been presented here so far, the use cases don't feel particularly strong (network service discovery and imap). Did I miss any others?

**Comment by ClaesNilsson** [comment]
Ok, thanks @jyasskin and @marcoscaceres. I will complete  #82.  

Regarding use cases we are interested in using UDP not only for multicast local network service discovery but also for getting sensor data via UDP. Marcos, you mention direct communication with IMAP servers. In addition in FFOS they are implementing UDP Sockets,  [FFOS UDP](https://bugzilla.mozilla.org/show_bug.cgi?id=745283) . Do you know anything about use cases here?

Google view on use cases? @domenic, @slightlyoff, @adrifelt,

**Comment by marcoscaceres** [comment]
So, for FXOS  "The motivating use case is DNS resolution for retrieval of MX records for the e-mail app's autoconfiguration mechanism." There are some people wanting to do home automation and service discovery stuff. Also, the proprietary API for FxOS already shipped - I'm not aware of plans to support this on the web tho (or to add this API to FxOS in addition to the one we already have).

**Comment by ClaesNilsson** [comment]
Sorry for delay but https://github.com/sysapps/tcp-udp-sockets/pull/82 is now updated to include UDPPermission and TCPServerpermission interfaces, https://github.com/ClaesNilsson/raw-sockets/commit/9d89c9c38ffbfc2a746afc969343f204e86ee797.

Any comments before I merge, @domenic, @slightlyoff, @jyasskin, @adrifelt, @marcoscaceres ?

**Comment by domenic** [comment]
Looks pretty reasonable to me, although my perspective is more micro- than macro-.

**Comment by ClaesNilsson** [comment]
Thanks Domenic, I see your perspective. So, as a side note, I have not been able to follow the latest discussion and updates to the Streams API. May I ask if you notice any changes needed to the descriptions of the UDP and TCPSocket's adaptation layer to the Streams API due to the recent changes done to Streams? My latest update on these parts are from October 27th. If so could you create a separate issue for that and I'll fix it.

**Comment by domenic** [comment]
@ClaesNilsson I think you are up to date on the latest streams (I have been trying not to change much), but I will do a review and open issues if I notice anything :).

**Comment by ClaesNilsson** [comment]
Merged https://github.com/sysapps/tcp-udp-sockets/pull/82

**Comment by ClaesNilsson** [comment]
Closed with https://github.com/sysapps/tcp-udp-sockets/pull/82
